### DISCUSSION TITLE
Azure openai not working: Unrecognized request argument supplied

### DISCUSSION THREAD

**Discussion by ghost** [567-labs/instructor#383] [Q&A] [answered]
I followed the example as such using azure openai and gpt4:
```import instructor
from openai import AzureOpenAI
from pydantic import BaseModel
 
# Enables `response_model`
client = instructor.patch(AzureOpenAI(
    api_version="2023-07-01-preview",
    azure_endpoint="https://<my_endpoint>.openai.azure.com",
))
class UserDetail(BaseModel):
    name: str
    age: int
 
user = client.chat.completions.create(
    model="gpt-4",
    response_model=UserDetail,
    messages=[
        {"role": "user", "content": "Extract Jason is 25 years old"},
    ]
)
 
print(user)
```
and got this error
```
openai.NotFoundError: Error code: 404 - {'error': {'message': 'Unrecognized request argument supplied: functions', 'type': 'invalid_request_error', 'param': None, 'code': None}}'
```

when i use the preview model:
```
import instructor
from openai import AzureOpenAI
from pydantic import BaseModel
 
# Enables `response_model`
client = instructor.patch(AzureOpenAI(
    api_version="2023-07-01-preview",
    azure_endpoint="https://<my_endpoint>.openai.azure.com",
))
class UserDetail(BaseModel):
    name: str
    age: int
 
user = client.chat.completions.create(
    model="gpt-4-1106-preview",
    response_model=UserDetail,
    messages=[
        {"role": "user", "content": "Extract Jason is 25 years old"},
    ]
)
 
print(user)
 ```
i get this error:
```
openai.NotFoundError: Error code: 404 - {'error': {'code': 'DeploymentNotFound', 'message': 'The API deployment for this resource does not exist. If you created the deployment within the last 5 minutes, please wait a moment and try again.'}}
```
I am new to instructor, if anyone could give any advice of this, thank you.

**Comment by JungeAlexander** [comment, marked as answer]
@hsu4 I just ran this code on my Azure GPT4-Turbo deployment and it works. My best guess as to why yours fails is that your GPT4 deployment is **not** called "gpt-4-1106-preview" but something else.

This is a parameter you can customize in Azure when first deploying the model and you can find the deployment name in Azure OpenAI Studio -> Management -> Deplyments as described [here](https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/create-resource?pivots=web-portal#deploy-a-model).

**Reply by ghost** [reply]
Thank you, I will check on this.

**Reply by matthewkperez** [reply]
I'm running into this same error. I have a model deployment setup which is a gpt-3.5-turbo model. 
I'm also using api_version="2023-09-01-preview" as suggested by this [discussion](https://github.com/jxnl/instructor/discussions/212#discussioncomment-7635760). However, I am still getting this error

openai.NotFoundError: Error code: 404 - {'error': {'message': 'Unrecognized request argument supplied: functions', 'type': 'invalid_request_error', 'param': None, 'code': None}}

```
def create_dynamic_model(input_string: str) -> BaseModel:
        words = input_string.split()

        # Creating a dictionary of field definitions
        fields = {word: (AllowedValue, Field(..., 
            description=f"The pharaphasia label for '{word}'. phonemic for rearrangment, \
            semantic for substitution, neologistic for nonsensical, and non-paraphaic for normal no error word."
        )) for word in words}

        # Creating a dynamic model using `create_model`
        DynamicModel = create_model('DynamicModel', **fields)

        return DynamicModel

transcript_labels = iclient.chat.completions.create(
            model="gpt-35-turbo-matt",
            response_model=create_dynamic_model(transcript_sentence),
            max_retries = 2,
            messages=[
                PROMPT,
                {
                    "role": "user",
                    "content": f"ASR Transcript: {transcript_sentence}",
                }
            ]
        )
```

**Reply by jxnl** [reply]
It means that your model does not support function calling. `gpt-35-turbo-matt`

**Reply by matthewkperez** [reply]
@jxnl Thanks for the reply. Is this something that can be toggled when the model is created on azure? Is this an issue with all azure models?

**Reply by matthewkperez** [reply]
I found this [resource](https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/function-calling?tabs=python#using-function-in-the-chat-completions-api-deprecated) that outlines which azure models have function call support

**Reply by jxnl** [reply]
Did the resource help answer the question?
